src/Api/InitializeInTracee.cpp

```diff
@@ -4,15 +4,101 @@
+using orbit_user_space_instrumentation::ExecuteInProcess;
+
+static ErrorMessageOr<absl::flat_hash_map<std::string, ModuleInfo>> GetModulesByPathForPid(
+    int32_t pid) {
+  OUTCOME_TRY(module_infos, orbit_elf_utils::ReadModules(pid));
+  absl::flat_hash_map<std::string, ModuleInfo> result;
+  for (const ModuleInfo& module_info : module_infos) {
+    result.emplace(module_info.file_path(), std::move(module_info));
+  }
+  return result;
+}
+
+[[nodiscard]] static const ModuleInfo* FindModuleInfoForApiFunction(
+    const ApiFunction& api_function,
+    const absl::flat_hash_map<std::string, ModuleInfo>& modules_by_path) {
+  auto module_info_it = modules_by_path.find(api_function.module_path());
+  if (module_info_it == modules_by_path.end()) {
+    ERROR("Could not find module \"%s\" when initializing Orbit Api.", api_function.module_path());
+    return nullptr;
+  }
+  const ModuleInfo& module_info = module_info_it->second;
+  if (module_info.build_id() != api_function.module_build_id()) {
+    ERROR("Build-id mismatch for \"%s\" when initializing Orbit Api", api_function.module_path());
+    return nullptr;
+  }
+
+  return &module_info;
+}
 
 namespace orbit_api {
+ErrorMessageOr<void> InitializeInTracee(const CaptureOptions& capture_options) {
+  if (capture_options.api_functions().size() == 0) {
+    return ErrorMessage("No api table to initialize.");
+  }
+
+  int32_t pid = capture_options.pid();
+
+  OUTCOME_TRY(AttachAndStopProcess(pid));
+
+  // Make sure we resume the target process, even on early-outs.
+  orbit_base::unique_resource scope_exit{pid, [](int32_t pid) {
+                                           if (DetachAndContinueProcess(pid).has_error()) {
+                                             ERROR("Detaching from %i", pid);
+                                           }
+                                         }};
+
+  // Load liborbit.so and find api table initialization function.
+  const std::string kLibPath = orbit_base::GetExecutableDir() / "liborbit.so";
+  constexpr const char* kInitFunction = "orbit_api_initialize";
+  OUTCOME_TRY(handle, DlopenInTracee(pid, kLibPath, RTLD_NOW));
+  OUTCOME_TRY(orbit_init_function, DlsymInTracee(pid, handle, kInitFunction));
+
+  // Initialize all api function tables.
+  OUTCOME_TRY(modules_by_path, GetModulesByPathForPid(pid));
+  for (const ApiFunction& api_function : capture_options.api_functions()) {
+    // Filter api functions.
+    constexpr const char* kOrbitApiGetAddressPrefix = "orbit_api_get_function_table_address_v";
+    if (absl::StrContains(api_function.name(), kOrbitApiGetAddressPrefix) == false) continue;
+
+    // Get ModuleInfo associated with function.
+    const ModuleInfo* module_info = FindModuleInfoForApiFunction(api_function, modules_by_path);
+    if (module_info == nullptr) continue;
+
+    // Get address of function table by calling "orbit_api_get_function_table_address_vN" in tracee.
+    void* api_function_address = absl::bit_cast<void*>(
```

Comment:
Nothing for this PR but I wonder if `void*` is the right choice to identify addresses in the trace. Both processes are 64 bit so it doesn't matter. I might consider to change the tracee side pointers to uint64_t in InjectLibraryInTracee.h and ExecuteInProcess.h.

Reply:
Yeah it's a good point, we should decide if we want to support 32 bit tracees or not, ideally we would, but then we need to add 32 bit versions for functions in InjectLibraryInTracee.h and ExecuteInProcess.h